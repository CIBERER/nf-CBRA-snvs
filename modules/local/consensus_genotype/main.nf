process CONSENSUS_GENOTYPE {
    tag "$meta.id"
    label 'process_low'

    conda "conda-forge::python=3.9.5 conda-forge::pandas=1.3.3"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/pandas:2.2.1':
        'biocontainers/pandas:2.2.1' }"

    input:
    tuple val(meta), path(gt_file)

    output:
    tuple val(meta), path("*_consensus_GT.txt"), emit: consensus_gt
    tuple val(meta), path("*_discordances.txt"), emit: discordances
    path "versions.yml"           , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    #!/usr/bin/env python3
    import pandas as pd
    import sys

    df = pd.read_table("${gt_file}", sep="\t")

    # Count NAs in each column and sort
    na_counts = df.isna().sum()
    na_counts = na_counts.sort_values(ascending=False)
    # Convert Series to DataFrame
    df_na_counts = na_counts.reset_index()
    df_na_counts.columns = ['Column', 'NA_Count']

    # Each caller has a different weight so that in case of tie, the order of prioritization goes from the one that got the most calls
    # Add a new column with values starting from 2 and increasing by 1
    df_na_counts['weight'] = range(2, 2 + len(df_na_counts))

    df_new = pd.DataFrame()

    for program in list(df.columns):
        df_weighted = pd.concat([df[program]] * list(df_na_counts[df_na_counts["Column"] == program]["weight"])[0], axis=1, ignore_index=True)
        df_new = pd.concat([df_new, df_weighted], axis = 1)

    # Write consensus Genotype
    df_new.mode(axis=1).to_csv("${prefix}_consensus_GT.txt", sep='\t', header=False, index=False)

    # Calculate and write discordances in genotype
    discordance_list = list(map(lambda x: len(set(filter(lambda y: y == y, x))) - 1, df.values))
    with open("${prefix}_discordances.txt", 'w') as fp:
        for item in discordance_list:
            fp.write(f"{item}\\n")

    with open("versions.yml", "w") as f:
        f.write('''
    "${task.process}":
        python: \$(python --version | sed 's/Python //g')
        pandas: \$(python -c "import pandas; print(pandas.__version__)")
    ''')
    """
}