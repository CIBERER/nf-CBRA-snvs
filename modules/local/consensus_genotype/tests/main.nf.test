nextflow_workflow {

    name "Test Subworkflow BCFTOOLS_QUERY_STATS"
    script "../main.nf"
    config "./nextflow.config"
    workflow "CONSENSUS_GENOTYPE"

    tag "modules"
    tag "modules/local/consensus_genotype"


   setup {

        run("BWA_INDEX") {
            script "../../../../modules/nf-core/bwa/index/main.nf"
            process {
                """
                input[0] = [
                    [ id: 'genome'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                ]
                """
            }
        }
        run("PICARD_CREATESEQUENCEDICTIONARY") {
            script "../../../../modules/nf-core/picard/createsequencedictionary/main.nf"
            process {
                """
                input[0] = [
                    [ id: 'genome'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                ]
                """
            }
        }
        run("MAPPING") {
            script "../../../../subworkflows/local/mapping/main.nf"
            process {
                """
                input[0] = Channel.of([
                    [ id : 'test_sample' ],
                    [file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/fastq/test.umi_1.fastq.gz', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/fastq/test.umi_2.fastq.gz', checkIfExists: true)]
                ])
                input[1] =  Channel.fromList([
                            tuple([ id: 'test_sample'],
                            [])
                        ])
                input[2] = BWA_INDEX.out.index
                input[3] =  Channel.fromList([
                        tuple([ id: 'genome'],
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true))
                    ])
                input[4] =  Channel.fromList([
                        tuple([ id: 'genome'],
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true))
                    ])
                input[5] = PICARD_CREATESEQUENCEDICTIONARY.out.reference_dict
                input[6] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/vcf/dbsnp_146.hg38.vcf.gz', checkIfExists: true)
                input[7] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/vcf/dbsnp_146.hg38.vcf.gz.tbi', checkIfExists: true)
                """
            }
        }

        run("GATK_VCF") {
            script "../../../../subworkflows/local/gatk_vcf/main.nf"
            process {
                """
                input[0] = MAPPING.out.bam
                input[1] =  Channel.fromList([
                        tuple([ id: 'test_sample'],
                        [])
                    ])
                input[2] =  Channel.fromList([
                        tuple([ id: 'genome'],
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true))
                    ])
                input[3] =  Channel.fromList([
                        tuple([ id: 'genome'],
                        file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true))
                    ])
                input[4] = PICARD_CREATESEQUENCEDICTIONARY.out.reference_dict
                input[5] = input[5] =  Channel.fromList([
                            tuple([ id: 'test_dbsnp'],
                            [])
                        ])
                input[6] = input[5] =  Channel.fromList([
                            tuple([ id: 'test_dbsnp'],
                            [])
                        ])
                """
            }
        }

        run("GATK4_COMPOSESTRTABLEFILE") {
            script "../../../../modules/nf-core/gatk4/composestrtablefile/main.nf"
            process {
                """
                input[0] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                input[1] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true)
                input[2] = file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.dict', checkIfExists: true)
                """
            }
        }

        run("DRAGEN_VCF") {
            script "../../../../subworkflows/local/dragen_vcf/main.nf"
            process {
                 """
                input[0] = MAPPING.out.bam
                input[1] = Channel.fromList([
                        tuple([ id: 'genome'],file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true))
                ])
                input[2] =  Channel.fromList([
                    tuple([ id: 'genome'],file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true))
                    ])
                input[3] = Channel.fromList([
                    tuple([ id: 'genome'],file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.dict', checkIfExists: true))
                ])
                input[4] = GATK4_COMPOSESTRTABLEFILE.out.str_table
                input[5] =  Channel.fromList([
                        tuple([ id: 'test_sample'],
                        [])
                    ])                
                input[6] = Channel.fromList([
                            tuple([ id: 'test_dbsnp'],
                            [])
                        ])
                input[7] = Channel.fromList([
                            tuple([ id: 'test_dbsnp'],
                            [])
                        ])                
                """
            }
        }


        run("BCFTOOLS_MERGE") {
            script "../../../../modules/nf-core/bcftools/merge/main.nf"
            process {
                 """
                input[0] = GATK_VCF.out.vcf.join(DRAGEN_VCF.out.vcf).map { item ->
                def meta = item[0]
                def files = item[1..-1].collect { file(it) }
                def vcfs = files.findAll { it.name.endsWith('.vcf.gz') }
                def tbis = files.findAll { it.name.endsWith('.vcf.gz.tbi') }
                [meta, vcfs, tbis]
                }
                input[1] = Channel.fromList([
                        tuple([ id: 'genome'],file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true))
                ])
                input[2] =  Channel.fromList([
                    tuple([ id: 'genome'],file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true))
                    ])
                input[3] =  Channel.fromList([
                    tuple([ id: 'test_sample'],
                    [])
                ]) 
         
                """
            }
        }

        run("TABIX_TABIX") {
            script "../../../../modules/nf-core/tabix/tabix/main.nf"
            process {
                 """
                input[0] = BCFTOOLS_MERGE.out.vcf
                """
            }
        }

        run("BCFTOOLS_QUERY_STATS") {
            script "../../../../modules/local/bcftools_query_stats/main.nf"
            process {
                 """
                input[0] = BCFTOOLS_MERGE.out.vcf.join(TABIX_TABIX.out.tbi)
                """
            }
        }

    }


    test("consensus_genotype - gt_file ") {


        when {
            workflow {
                """
                input[0] = BCFTOOLS_QUERY_STATS.out.gt
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }

}
