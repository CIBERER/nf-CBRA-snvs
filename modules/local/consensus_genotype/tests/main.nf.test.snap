{
    "consensus_genotype - gt_file ": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_consensus_GT.txt:md5,cb97fbabc44fdcd264133aa00ca06c97"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_discordances.txt:md5,e341ed49af2d21d764eda0f46a63293a"
                    ]
                ],
                "2": [
                    "versions.yml:md5,12f63a12e6f670cd4d627ff4d9f704de"
                ],
                "consensus_gt": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_consensus_GT.txt:md5,cb97fbabc44fdcd264133aa00ca06c97"
                    ]
                ],
                "discordances": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_discordances.txt:md5,e341ed49af2d21d764eda0f46a63293a"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,12f63a12e6f670cd4d627ff4d9f704de"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.04.4"
        },
        "timestamp": "2025-05-05T10:02:25.400025608"
    }
}